File: lib/gemini.ts
export async function generateContent(prompt: string, options: {
  maxTokens?: number;
  temperature?: number;
  topK?: number;
  topP?: number;
} = {}) {
  const { maxTokens = 1500, temperature = 0.8, topK = 40, topP = 0.95 } = options;

  if (!process.env.GEMINI_API_KEY) {
    console.error("Missing GEMINI_API_KEY");
    throw new Error("Server configuration error");
  }

  let attempt = 0;
  const maxRetries = 3;
  let response;
  let data;

  while (attempt <= maxRetries) {
    try {
      if (attempt > 0) {
        const delay = Math.pow(2, attempt) * 1000; // 2s, 4s, 8s
        console.log(`⚠️ Gemini API overloaded/rate-limited. Retrying in ${delay}ms (Attempt ${attempt}/${maxRetries})...`);
        await new Promise(resolve => setTimeout(resolve, delay));
      }

      console.log(`Sending Gemini API request (Attempt ${attempt + 1})...`);
      response = await fetch(
        `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-pro:generateContent?key=${process.env.GEMINI_API_KEY}`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { temperature, topK, topP, maxOutputTokens: maxTokens },
          }),
        }
      );

      console.log('Gemini API response status:', response.status);
      data = await response.json();

      if (response.ok) {
        break; // Success!
      }

      // Handle retryable errors
      if (response.status === 503 || response.status === 429) {
        attempt++;
        continue;
      }

      // Fatal error
      console.error('Gemini API error response:', JSON.stringify(data, null, 2));
      throw new Error(`Gemini API error: ${response.status} - ${JSON.stringify(data)}`);

    } catch (networkError) {
      // Network errors (like fetch failing) might also be worth retrying
      console.error(`Attempt ${attempt + 1} failed network error:`, networkError);
      if (attempt < maxRetries) {
        attempt++;
        continue;
      }
      throw networkError;
    }
  }

  if (!response || !response.ok) {
    throw new Error('Max retries exceeded for Gemini API');
  }

  // data has already been parsed above
  // console.log('Gemini API full response:', JSON.stringify(data, null, 2)); // Verbose logging removed for clarity

  if (!data.candidates || data.candidates.length === 0) {
    throw new Error(`No candidates returned from Gemini. Response: ${JSON.stringify(data, null, 2)}`);
  }

  if (!data.candidates[0].content?.parts?.[0]?.text) {
    throw new Error(`No content in candidates. Response: ${JSON.stringify(data, null, 2)}`);
  }

  const content = data.candidates[0].content.parts[0].text.trim();
  console.log('Gemini API generated content length:', content.length);
  console.log('Token usage:', JSON.stringify(data.usageMetadata, null, 2));
  return content;
}

File: python-backend/app/services/gemini.py
"""Gemini API client for generating automation content.

Mirrors the logic in lib/gemini.ts with retry support.
"""

import httpx
import asyncio
import json
import logging

from app.config import get_settings

logger = logging.getLogger(__name__)


async def generate_content(
    prompt: str,
    max_tokens: int = 3000,
    temperature: float = 0.8,
    top_k: int = 40,
    top_p: float = 0.95,
) -> str:
    """Generate content using Gemini API with retry logic."""
    settings = get_settings()

    if not settings.gemini_api_key:
        raise ValueError("GEMINI_API_KEY is not set")

    url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key={settings.gemini_api_key}"

    body = {
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": {
            "temperature": temperature,
            "topK": top_k,
            "topP": top_p,
            "maxOutputTokens": max_tokens,
        },
    }

    max_retries = 3
    for attempt in range(max_retries + 1):
        try:
            if attempt > 0:
                delay = 2 ** attempt
                logger.warning(f"Gemini retry {attempt}/{max_retries}, waiting {delay}s...")
                await asyncio.sleep(delay)

            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    url,
                    json=body,
                    headers={"Content-Type": "application/json"},
                )

                if response.status_code == 200:
                    data = response.json()
                    candidates = data.get("candidates", [])
                    if not candidates:
                        raise ValueError("No candidates returned from Gemini")

                    text = candidates[0].get("content", {}).get("parts", [{}])[0].get("text", "")
                    if not text:
                        raise ValueError("No text content in Gemini response")

                    return text.strip()

                if response.status_code in (429, 503):
                    continue  # Retry

                raise ValueError(f"Gemini API error {response.status_code}: {response.text}")

        except httpx.HTTPError as e:
            if attempt < max_retries:
                continue
            raise

    raise ValueError("Max retries exceeded for Gemini API")


def parse_json_response(raw: str) -> list[dict]:
    """Parse JSON from Gemini response, handling code fences and repairs."""
    # Remove code fences
    cleaned = raw.replace("```json\n", "").replace("\n```", "").replace("```", "").strip()

    try:
        result = json.loads(cleaned)
        if isinstance(result, list):
            return result
        return [result]
    except json.JSONDecodeError:
        pass

    # Try extracting JSON array
    start = cleaned.find("[")
    end = cleaned.rfind("]")
    if start != -1 and end != -1 and end > start:
        try:
            return json.loads(cleaned[start : end + 1])
        except json.JSONDecodeError:
            pass

    # Soft repair: try adding closing bracket
    if "[" in cleaned and "]" not in cleaned:
        cleaned = cleaned.rstrip(",") + "]"
        try:
            return json.loads(cleaned[cleaned.find("[") :])
        except json.JSONDecodeError:
            pass

    raise ValueError(f"Failed to parse Gemini response as JSON: {raw[:200]}...")
